specifyweb/specify/import_locality.py

```diff
@@ -0,0 +1,488 @@
+                'taskid': li.taskid,
+                "localitites": localitites,
+                "geocoorddetails": geocoorddetails
+            }))
+
+    # prevent Celery from overriding the State of the Task
+    raise Ignore()
+
+
+class JSONParseError(TypedDict):
+    message: str
+    field: str
+    payload: Dict[str, Any]
+    rowNumber: int
+
+
+class ParseError(NamedTuple):
+    message: Union[ParseFailureKey, LocalityParseErrorMessageKey]
+    field: Optional[str]
+    payload: Optional[Dict[str, Any]]
+    row_number: Optional[int]
+
+    @classmethod
+    def from_parse_failure(cls, parse_failure: BaseParseFailure, field: str, row_number: int):
+        return cls(message=parse_failure.message, field=field, payload=parse_failure.payload, row_number=row_number)
+
+    def to_json(self) -> JSONParseError:
+        return {"message": self.message, "field": self.field, "payload": self.payload, "rowNumber": self.row_number}
+
+
+class ParsedRow(TypedDict):
+    row_number: int
+    locality: Dict[str, Any]
+    geocoorddetail: Optional[Dict[str, Any]]
+    locality_id: int
+
+
+class ParseSuccess(NamedTuple):
+    to_upload: Dict[str, Any]
+    model: ImportModel
+    locality_id: Optional[int]
+    row_number: Optional[str]
+
+    @classmethod
+    def from_base_parse_success(cls, parse_success: BaseParseSuccess, model: ImportModel, locality_id: Optional[int], row_number: int):
+        return cls(parse_success.to_upload, model, locality_id, row_number)
+
+
+class UploadSuccessRow(TypedDict):
+    locality: int
+    geocoorddetail: Optional[int]
+
+
+class UploadSuccess(TypedDict):
+    type: Literal["Uploaded"]
+    results: List[UploadSuccessRow]
+
+
+class UploadParseError(TypedDict):
+    type: Literal["ParseError"]
+    errors: List[ParseError]
+
+
+@transaction.atomic
+def resolve_localityimport_result(taskid: str, results: Union[Tuple[List[ParsedRow], List[ParseError]], Union[UploadSuccess, UploadParseError]], collection, create_recordset: bool = False) -> LocalityImport:
+
+    li = LocalityImport.objects.get(taskid=taskid)
+
+    # the results come from parse_locality_set
+    if isinstance(results, tuple):
+        to_upload, errors = results
+        if len(errors) > 0:
+            status = LocalityImportStatus.FAILED
+            for error in errors:
+                result = error.to_json()
+                LocalityImportRowResult.objects.create(
+                    localityimport=li,
+                    rownumber=result["rowNumber"],
+                    result=json.dumps(result, cls=DjangoJSONEncoder)
+                )
+        else:
+            status = LocalityImportStatus.PARSED
+            localities = len(to_upload)
+            geocoorddetails = 0
+            for parsed in to_upload:
+                if parsed['geocoorddetail'] is not None:
+                    geocoorddetails += 1
+            LocalityImportRowResult.objects.create(
+                localityimport=li,
+                rownumber=-1,
```

Comment:
Simplify some things here. Why `-1`? that's hacky and just bad. What if we want to add a feature that allows users to parse the dataset, and then open the matched localities in a new tab? Then, we need to have the ids of the localities. your current code blocks it. what's worse is that if we add this in the future, then we need to handle the ugly case that there is _some_ dataset out there with `-1`.

Suggested refactor. Do this incrementally.

**Get rid of the state "ParseError" and "Uploaded".** 

Yes, I mean even from upload call. Instead, _infer_ it later on. A very good reason for this is what if there is a business rule or an exception that happens at the DB level? Your code doesn't have any capacity to handle that. And, it is not even that hard to support it. + this means there is no concept of ParseError and Uploaded, so the entire second branch of this function isn't needed. Instead, just make results a `List[Union[UploadedResult, ParsedResult]`. Here, a row can either be parsed or uploaded. Actually, it is almost similar to what you have with `ParsedRow`. Just make add another key `errors`. Change your `parse_locality_set` to do so. That is, parse_locality_set will return `List[ParsedResult]`

```
class ParsedBase(NamedTuple):
    row_number: Optional[int]
    locality: Dict[str, Any]
    geocoorddetail: Optional[Dict[str, Any]]

class ParsedSuccess(ParsedBase):
    locality_id: int

class ParsedFail(ParseBase):
    errors: List[ParseError] 

ParsedRow = ParsedFail | ParsedSuccess

```

Right before upload, check if any row is ParsedFail. If so, skip the upload. Otherwise, on upload success, it will be `UploadSuccessRow`.

So, the entire results will be `List[Union[ParsedFail, Union[ParsedSuccess, UploadSuccessRow]]]`

The benefit of this approach is that the entire `isinstance(results, tuple)` can be removed (can be combined with the later branch). In the future, if we want to include business rules, then this will be trivial (convert failure to ParsedFail). 

Will also simplify 
```
        if results['type'] == 'ParseError':
            status = LocalityImportStatus.FAILED
            for error in results['errors']:
                result = error.to_json()
                LocalityImportRowResult.objects.create(
                    localityimport=li,
                    rownumber=error.row_number,
                    result=json.dumps(result, cls=DjangoJSONEncoder)
                )

        elif results['type'] == 'Uploaded':
            status = LocalityImportStatus.SUCCEEDED
            localities = []
            for index, row in enumerate(results['results']):
                row_number = index + 1
                localities.append(row['locality'])

                LocalityImportRowResult.objects.create(
                    localityimport=li,
                    rownumber=row_number,
                    result=json.dumps(row, cls=DjangoJSONEncoder)
                )
```
to be
```
 for index, row in enumerate(results):
            LocalityImportRowResult.objects.create(
                    localityimport=li,
                    rownumber=row_number,
                    result=json.dumps(row, cls=DjangoJSONEncoder)
                )
            if isinstance(row, UploadSuccessRow):
               localities.append(row['locality'])
```

Reply:
The rownumber issue has been addressed, but the refactoring of `parse_locality_set` may need to wait as this PR is holding up the release. 
I will open an issue for this.